receiver/carbonreceiver/protocol/regex_parser.go

```diff
@@ -0,0 +1,209 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package protocol
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+)
+
+const (
+	metricNameCapturePrefix = "name_part"
+)
+
+// RegexParserConfig has the configuration for a parser that can breakdown a
+// Carbon "metric path" and transform it in corresponding metric labels according
+// to a series of regular expressions rules (see below for details).
+//
+// This is typically used to extract labels from a "naming hierarchy", see
+// https://graphite.readthedocs.io/en/latest/feeding-carbon.html#step-1-plan-a-naming-hierarchy
+//
+// Examples:
+//
+// 1. Rule:
+//        - regexp: "(?P<svc>[^.]+)\.(?P<host>[^.]+)\.cpu\.seconds"
+//          name_prefix: cpu_seconds
+//          labels:
+//            k: v
+//    Metric path: "service_name.host00.cpu.seconds"
+//    Resulting metric:
+//        name: cpu_seconds
+//        label keys: {"svc", "host", "k"}
+//        label values: {"service_name", "host00", "k"}
+//
+// 2. Rule:
+//        - regexp: "^(?P<svc>[^.]+)\.(?P<host>[^.]+)\.(?P<name_part0>[^.]+).(?P<name_part1>[^.]+)$"
+//    Metric path: "svc_02.host02.avg.duration"
+//    Resulting metric:
+//        name: avgduration
+//        label keys: {"svc", "host"}
+//        label values: {"svc_02", "host02"}
+//
+type RegexParserConfig struct {
+	// Rules contains the regular expression rules to be used by the parser.
+	// The first rule that matches and applies the transformations configured in
+	// the respective RegexRule struct. If no rules match the metric is then
+	// processed by the "plaintext" parser.
+	Rules []*RegexRule `mapstructure:"rules"`
+
+	// MetricNameSeparator is used when joining the name prefix of each individual
+	// rule and the respective named captures that start with the prefix
+	// "name_part" (see RegexRule for more information).
+	MetricNameSeparator string `mapstructure:"name_separator"`
+}
+
+// RegexRule describes how parts of the name of metric are going to be mapped
+// to metric labels. The rule is only applied if the name matches the given
+// regular expression.
+type RegexRule struct {
+	// Regular expression from which named matches are used to extract label
+	// keys and values from Carbon metric paths.
+	Regexp string `mapstrucutre:"regexp"`
+
+	// NamePrefix is the prefix added to the metric name after extracting the
+	// parts that will form labels and final metric name.
+	NamePrefix string `mapstructure:"name_prefix"`
+
+	// Labels are key-value pairs added as labels to the metrics that match this
+	// rule.
+	Labels map[string]string `mapstructure:"labels"`
+
+	// Counter forces the metric to be a counter, by default all metrics are set
+	// to be gauges this flag can be used to make it a cumulative counter.
+	Counter bool `mapstructure:"counter"`
```

Comment:
Are "gauge" and "counter" the only possible metric types that may ever be supported here? If any other type is expected to be added it may be worth to use an string/enum instead of bool, e.g. "type: counter" in instead of "counter: true" and make "type: gauge" the default.